src/Client/ClientBase.cpp

```diff
@@ -0,0 +1,1367 @@
+    {
+        if (!insert->format.empty())
+            current_format = insert->format;
+    }
+
+    auto source = FormatFactory::instance().getInput(current_format, buf, sample, global_context, insert_format_max_block_size);
+    Pipe pipe(source);
+
+    if (columns_description.hasDefaults())
+    {
+        pipe.addSimpleTransform([&](const Block & header)
+        {
+            return std::make_shared<AddingDefaultsTransform>(header, columns_description, *source, global_context);
+        });
+    }
+
+    QueryPipeline pipeline;
+    pipeline.init(std::move(pipe));
+    PullingAsyncPipelineExecutor executor(pipeline);
+
+    Block block;
+    while (executor.pull(block))
+    {
+        /// Check if server send Log packet
+        receiveLogs(parsed_query);
+
+        /// Check if server send Exception packet
+        auto packet_type = connection->checkPacket();
+        if (packet_type && *packet_type == Protocol::Server::Exception)
+        {
+            /*
+                * We're exiting with error, so it makes sense to kill the
+                * input stream without waiting for it to complete.
+                */
+            executor.cancel();
+            return;
+        }
+
+        if (block)
+        {
+            connection->sendData(block, /* name */"", /* scalar */false);
+            processed_rows += block.rows();
+        }
+    }
+
+    connection->sendData({}, "", false);
+}
+
+
+/// Process Log packets, used when inserting data by blocks
+void ClientBase::receiveLogs(ASTPtr parsed_query)
+{
+    auto packet_type = connection->checkPacket(0);
+
+    while (packet_type && *packet_type == Protocol::Server::Log)
+    {
+        receiveAndProcessPacket(parsed_query, false);
+        packet_type = connection->checkPacket();
+    }
+}
+
+
+/// Process Log packets, exit when receive Exception or EndOfStream
+bool ClientBase::receiveEndOfQuery()
+{
+    while (true)
+    {
+        Packet packet = connection->receivePacket();
+
+        switch (packet.type)
+        {
+            case Protocol::Server::EndOfStream:
+                onEndOfStream();
+                return true;
+
+            case Protocol::Server::Exception:
+                onReceiveExceptionFromServer(std::move(packet.exception));
+                return false;
+
+            case Protocol::Server::Log:
+                onLogData(packet.block);
+                break;
+
+            default:
+                throw NetException(
+                    "Unexpected packet from server (expected Exception, EndOfStream or Log, got "
+                        + String(Protocol::Server::toString(packet.type)) + ")",
+                    ErrorCodes::UNEXPECTED_PACKET_FROM_SERVER);
+        }
+    }
+}
+
+
+void ClientBase::processParsedSingleQuery(const String & full_query, const String & query_to_execute,
+        ASTPtr parsed_query, std::optional<bool> echo_query_, bool report_error)
+{
+    resetOutput();
+    have_error = false;
```

Comment:
It is done separately, because local non-interactive has a different from client behaviour with handling exceptions (like which exceptions and when are rethrown).

Reply:
It's easy to forget to reset something when it's done in different places

Reply:
Now localServer non-intaractive if `ignore_error == true` will anyway throw but only when all queries are processed: 
https://github.com/ClickHouse/ClickHouse/blob/311110e36b55577e6d4c3cffb8d43eb02e52ccf8/programs/local/LocalServer.cpp#L450-L461
It seems strange, may be remove this behaviour? (but it will require to adjust some tests)
Then there will also be no need to deal with exceptions separately for client and local.

Reply:
I think it's ok to change this behavior and make it consistent with clickhouse-client.